### ISSUE TITLE
Добавление датчиков SCD4x

### ISSUE THREAD

**Issue by kamulchik** [Pshatsillo/mega_hacs#3]
**Описание**
Прошу добавить поддержку датчиков I2C SCD4x, например SCD41.
Вывод датчика:
CO2:919/temp:21.73/hum:49.71

**Comment by Pshatsillo** [comment]
Готово

**Comment by kamulchik** [comment]
Пока не могу согласиться... в устройстве нет объектов:
![image](https://github.com/user-attachments/assets/8c478536-0159-49c0-acf0-911bc7493dda)

Настройки порта на меге:
![image](https://github.com/user-attachments/assets/f7942cd5-04d4-4152-aa2b-3173b853431c)

**Comment by Pshatsillo** [comment]
reload objects делали?

**Comment by kamulchik** [comment]
Да, конечно, вот лог после нажатия кнопки "обновить объекты". Как вариант, раньше порт был в skip true, возможно он не вернулся из этого состояния...
`2024-11-26 17:32:33.216 DEBUG (MainThread) [custom_components.mega.mega3] find scan link: None
2024-11-26 17:32:33.216 DEBUG (MainThread) [custom_components.mega.mega3] request: http://192.168.1.13/sec/?pt=39
2024-11-26 17:32:33.246 DEBUG (MainThread) [custom_components.mega.mega3] response <meta charset=windows-1251><style>input,select{margin:1px}</style><a href=/sec>Back</a><br>P39<br>CO2:1060/temp:20.93/hum:50.78<form action=/sec/><input type=hidden name=pn value=39>Type <select name=pty><option value=255>NC<option value=0>In<option value=1>Out<option value=3>DSen<option value=4 selected>I2C</select><br>Mode <select name=m><option value=0>NC<option value=1 selected>SDA<option value=2>SCL</select><br>SCL <input name=misc size=3 value=38><br>Cat <select name=gr><option value=0>ANY<option value=1>Temp/Hum<option value=2>Light<option value=3>Expander<option value=5 selected>Air Quality<option value=4>Misc</select><br>Dev <select name=d><option value=0>NC<option value=40>T67xx<option value=44 selected>SCD4x<option value=41>HM3301<option value=42>SPS30</select><br><a href=/sec/?pt=39&cmd=scan>I2C Scan</a><br>Title <input name=emt size=25 value="scd_co2"><br><input type=submit value=Save></form>
2024-11-26 17:32:33.259 DEBUG (MainThread) [custom_components.mega.mega3] find scan link: <a href="/sec/?pt=39&amp;cmd=scan">I2C Scan</a>
2024-11-26 17:32:33.259 DEBUG (MainThread) [custom_components.mega.mega3] request: http://192.168.1.13/sec/?pt=39&cmd=scan
2024-11-26 17:32:33.369 DEBUG (MainThread) [custom_components.mega.mega3] response <a href=/sec/?pt=39>Back</a><br>0x62 - <a href=/sec/?pt=39&scl=38&i2c_dev=scd4x>SCD4x</a><br>
2024-11-26 17:32:33.372 DEBUG (MainThread) [custom_components.mega.mega3] scan results: ([{'pt': '39', 'scl': '38', 'i2c_dev': 'scd4x', 'i2c_par': 1}, {'pt': '39', 'scl': '38', 'i2c_dev': 'scd4x', 'i2c_par': 2}, {'pt': '39', 'scl': '38', 'i2c_dev': 'scd4x', 'i2c_par': 3}], [{'id_suffix': 'scd4x_carbon_dioxide', 'device_class': <SensorDeviceClass.CO2: 'carbon_dioxide'>, 'params': {'pt': '39', 'scl': '38', 'i2c_dev': 'scd4x', 'i2c_par': 1}, 'unit_of_measurement': 'ppm'}, {'id_suffix': 'scd4x_temperature', 'device_class': <SensorDeviceClass.TEMPERATURE: 'temperature'>, 'params': {'pt': '39', 'scl': '38', 'i2c_dev': 'scd4x', 'i2c_par': 2}, 'unit_of_measurement': <UnitOfTemperature.CELSIUS: '°C'>}, {'id_suffix': 'scd4x_humidity', 'device_class': <SensorDeviceClass.HUMIDITY: 'humidity'>, 'params': {'pt': '39', 'scl': '38', 'i2c_dev': 'scd4x', 'i2c_par': 3}, 'unit_of_measurement': '%'}])

2024-11-26 17:32:53.653 DEBUG (MainThread) [custom_components.mega.mega3] subscribe 37
2024-11-26 17:32:53.654 DEBUG (MainThread) [custom_components.mega.entities] could not convert to float "None"
2024-11-26 17:32:53.655 DEBUG (MainThread) [custom_components.mega.mega3] poll
2024-11-26 17:32:53.655 DEBUG (MainThread) [custom_components.mega.mega3] request: http://192.168.1.13/sec/?pt=39&scl=38&i2c_dev=scd4x&i2c_par=1
2024-11-26 17:32:53.683 DEBUG (MainThread) [custom_components.mega.mega3] response 1060
2024-11-26 17:32:53.684 DEBUG (MainThread) [custom_components.mega.mega3] i2c response: {(('pt', '39'), ('scl', '38'), ('i2c_dev', 'scd4x'), ('i2c_par', 1)): '1060'}
2024-11-26 17:32:53.684 DEBUG (MainThread) [custom_components.mega.mega3] request: http://192.168.1.13/sec/?pt=39&scl=38&i2c_dev=scd4x&i2c_par=2
2024-11-26 17:32:53.702 DEBUG (MainThread) [custom_components.mega.mega3] response 20.91
2024-11-26 17:32:53.703 DEBUG (MainThread) [custom_components.mega.mega3] i2c response: {(('pt', '39'), ('scl', '38'), ('i2c_dev', 'scd4x'), ('i2c_par', 2)): '20.91'}
2024-11-26 17:32:53.703 DEBUG (MainThread) [custom_components.mega.mega3] request: http://192.168.1.13/sec/?pt=39&scl=38&i2c_dev=scd4x&i2c_par=3
2024-11-26 17:32:53.720 DEBUG (MainThread) [custom_components.mega.mega3] response 50.78
2024-11-26 17:32:53.721 DEBUG (MainThread) [custom_components.mega.mega3] i2c response: {(('pt', '39'), ('scl', '38'), ('i2c_dev', 'scd4x'), ('i2c_par', 3)): '50.78'}
2024-11-26 17:32:53.722 DEBUG (MainThread) [custom_components.mega.mega3] request: http://192.168.1.13/sec/?pt=40&scl=41&i2c_dev=max44009
2024-11-26 17:32:53.737 DEBUG (MainThread) [custom_components.mega.mega3] response 4.14
2024-11-26 17:32:53.739 DEBUG (MainThread) [custom_components.mega.mega3] i2c response: {(('pt', '40'), ('scl', '41'), ('i2c_dev', 'max44009')): '4.14'}
2024-11-26 17:32:53.739 DEBUG (MainThread) [custom_components.mega.mega3] request: http://192.168.1.13/sec/?pt=40&scl=41&i2c_dev=bmp180
2024-11-26 17:32:53.771 DEBUG (MainThread) [custom_components.mega.mega3] response 4294965856.4294967282
`

**Comment by Pshatsillo** [comment]
Попробуйте несколько раз обновить объекты. Есть проблема, что если датчик
не отрабатывает, то объект не создаётся, как раз none на это косвенно
намекает. Дальше обновляться будет нормально. К сожалению создание энтити
завязано на сканирование меги. И изменить это в данной реализации
невозможно

вт, 26 нояб. 2024 г., 20:39 kamulchik ***@***.***>:

**Comment by devsurgaz** [comment]
> Не помогло. None отдаёт 37 порт, он неинтересен. Нужный порт всё отдаёт по логу: 'pt': '39', 'scl': '38' И адрес в том числе: response <a href=/sec/?pt=39>Back 0x62 - <a href=/sec/?pt=39&scl=38&i2c_dev=scd4x>SCD4x
Просто ради эксперимента - попробуйте перезагрузить сам контроллер, а затем сделать reload objects. У меня объекты создаются при перезагрузке меги, а опрос объектов начинается после "reload objects" в интеграции. Пока не смог победить этот баг

**Comment by Pshatsillo** [comment]
Тут выяснилась интересная штука: а вы прошивку контроллера до свежих версий
не обновляли? Со старыми прошивками работать не будет

чт, 5 дек. 2024 г., 11:02 devsurgaz ***@***.***>:

**Comment by kamulchik** [comment]
Прошивка меги: 4.63b8
_попробуйте перезагрузить сам контроллер, а затем сделать reload objects_
Не помогло.
_Возможно действительно что-то с инициализацией порта?_
Очень похоже. Порт перевёл в NC, он же тогда не должен быть устройством в HA? Но он остался после обновления объектов и перезагрузки HA... Т.е. устройство без объектов...
И пока не придумал как это дебажить...

**Comment by kamulchik** [comment]
Да, добавил этот же контроллер, но с другим адресом - всё работает, объекты подтягиваются... Но пока это сложное решение удалить контроллер из HA и добавить заново... много переписывать...
Из сложных устройств, которых давно нет на контроллере и в конфиге HA - led лента WS2812. Проблема как-то связана с ней...

**Comment by Pshatsillo** [comment]
Возможно какое-то действие вызвало ошибку в удалении конфига и теперь он не может очистить сам себя из БД

**Comment by kamulchik** [comment]
Решено удалением и добавлением контроллера
